mlflow/tensorflow.py

```diff
@@ -657,34 +657,82 @@ def autolog(every_n_iter=100):
                       "1.12 <= v <= 2.0.0 are supported.")
         return
 
+    @gorilla.patch(tensorflow.estimator.Estimator)
+    def train(self, *args, **kwargs):
+        if not mlflow.active_run():
+            try_mlflow_log(mlflow.start_run)
+            global _AUTOLOG_RUN_ID
+            _AUTOLOG_RUN_ID = mlflow.active_run().info.run_id
+
+        original = gorilla.get_original_attribute(tensorflow.estimator.Estimator, 'train')
+
+        # Checking step and max_step parameters for logging
+        if len(args) >= 3:
+            try_mlflow_log(mlflow.log_param, 'steps', args[2])
+            if len(args) >= 4:
+                try_mlflow_log(mlflow.log_param, 'max_steps', args[3])
+        if 'steps' in kwargs:
+            try_mlflow_log(mlflow.log_param, 'steps', kwargs['steps'])
+        if 'max_steps' in kwargs:
+            try_mlflow_log(mlflow.log_param, 'max_steps', kwargs['max_steps'])
+
+        result = original(self, *args, **kwargs)
+
+        if mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID:
```

Comment:
```suggestion
        if mlflow.active_run() is not None and mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID
```

Just to be safe. Also, we could pull this into a helper as well (e.g. `_maybe_end_active_run`) to reduce duplication

Reply:
Actually, you could wrap both the run-creation & run-ending logic in a `contextmanager`, e.g.

```
    @contextmanager
    def _manage_active_run():
        if not mlflow.active_run():
            try_mlflow_log(mlflow.start_run)
        yield mlflow.active_run()
        if mlflow.active_run() is not None and mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID:
            try_mlflow_log(mlflow.end_run)        
```

and then use it via `with _manage_active_run(): ... # do stuff in with block`

https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager has more info on this pattern, we also use it in https://github.com/mlflow/mlflow/blob/690c7e933f4c1007f477fa2b240631a9f8f515cf/mlflow/store/db/utils.py#L68

Reply:
Looks good to me, I just changed the function a little to the following:

```
    @contextmanager
    def _manage_active_run():
        if not mlflow.active_run():
            try_mlflow_log(mlflow.start_run)
            global _AUTOLOG_RUN_ID
            if mlflow.active_run() is not None: # defensive check in case `mlflow.start_run` fails
                _AUTOLOG_RUN_ID = mlflow.active_run().info.run_id
        yield mlflow.active_run()
        if mlflow.active_run() is not None and mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID:
            try_mlflow_log(mlflow.end_run)  
```
This is so `_AUTOLOG_RUN_ID` wouldn't be set unless the run is created by the first `if` statement.